repo2docker/contentproviders/dataverse.py

```diff
@@ -129,4 +213,4 @@ def fetch(self, spec, output_dir, yield_output=False):
     @property
     def content_id(self):
         """The Dataverse persistent identifier."""
-        return self.record_id
+        return hashlib.sha256(self.url.encode()).hexdigest()
```

Comment:
Is there a risk that a 64 character hash might make image names too long?

Reply:
https://github.com/opencontainers/distribution-spec/blob/main/spec.md#pulling-manifests says many implementations limit the hostname and name of the image in total to 256 chars. I *think* this means it may be good enough and not a problem?

Alternatively, I can go back to parsing persistent_id at `detect` time, instead of at `fetch` time, and set it that way. I think part of the confusion here is around `detect` semantics and when `content_id` is called. Ideally `detect` should be stateless and be simply used to, well, detect things! But we seem to treat it as also the thing that sets `.content_id` so it's a little bit of a mess. I'm happy to treat that as a different refactor though.

## Choice to be made

1. Leave this as is
2. Set this to be persistent_id instead, and move persistent_id parsing back into `detect`

Happy to do either!

Reply:
> But we seem to treat it as also the thing that sets .content_id so it's a little bit of a mess.

Yeah, that certainly doesn't _sound_ right. It looks to me like we also only access content_id [after calling fetch](https://github.com/jupyterhub/repo2docker/blob/60c0d70f373fc1ec1ee9542709757b681717e8ac/repo2docker/app.py#L507-L520). Is it possible that the issue you are seeing is only in the tests, not how r2d actually behaves? What happens if you `raise` in content_id if fetch hasn't been called?

If it's just the tests and persistent_id is defined after fetch, then keeping persistent_id seems nice here, and maybe we can fix the tests to be more realistic. And make it _explicit_ that `content_id` cannot be assumed to be available until `fetch` has been called?


----

A tangent I went on about hash length, that I'm not sure is relevant anymore, but already wrote down. Feel free to ignore:

Initially, I thought the content id was the _full_ thing, but of course it's the 'ref' that goes after the doi itself. Running a test gives this 106-character image name:

```
r2ddoi-3a10-2e7910-2fdvn-2f6zxagt-2f3yrryjec19b07b80bf8eeb95f669a51f64efb7f647f91cf1b1f6ccbef736396ba936ef
```

Since we're in the namespace of the doi, collision probability is _super_ low. We [truncate to the short hash in git](https://github.com/jupyterhub/repo2docker/blob/60c0d70f373fc1ec1ee9542709757b681717e8ac/repo2docker/contentproviders/git.py#L81). So maybe truncate this hash, or use a natively shorter hash function like:

```python
hashlib.blake2s(self.url.encode(), digest_size=10).hexdigest()
```

(blake2 is in `hashlib.algorithms_guaranteed` since 3.6, I think)

Reply:
ooooh, fixing the tests seems the right thing to do! I'll take a peek.

Reply:
hey, it looks like I already fixed the tests, so it's all fine now! Back to using persistent_id as the identifier, but this time it's 'proper' - so if we get a file persistent_id, we resolve it to the dataset persistent id, and use *that*. so if multiple different folks try to use different files from the same dataset, it will lead to cache reuse now!

Reply:
Great! All looks right to me, then.